### ISSUE TITLE
Loss goes to 0, Gibberish Outputs

### ISSUE THREAD

**Issue by radna0** [modelscope/ms-swift#3582]
```
swift rlhf \
    --rlhf_type grpo \
    --model /storage/model \
    --model_type deepseek_r1_distill \
    --use_hf true \
    --async_generate false \
    --num_infer_workers 8 \
    --num_iterations 2 \
    --reward_funcs cosine \
    --cosine_max_len 12288 \
    --cosine_max_len_value_correct 0.4 \
    --cosine_min_len_value_correct 0.8 \
    --cosine_max_len_value_wrong 0 \
    --cosine_min_len_value_wrong -0.2 \
    --train_type lora \
    --lora_rank 32 \
    --lora_alpha 32 \
    --target_modules all-linear \
    --torch_dtype bfloat16 \
    --dataset dataset \
    --max_completion_length 12288 \
    --num_train_epochs 5 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 4 \
    --eval_steps 8 \
    --save_steps 8 \
    --save_total_limit 100 \
    --logging_steps 1 \
    --report_to wandb \
    --max_length 16384 \
    --output_dir /storage/trained_grpo \
    --warmup_ratio 0.1 \
    --num_generations 4 \
    --temperature 1.0 \
    --top_p 0.90 \
    --top_k 50 \
    --repetition_penalty 1.1 \
    --deepspeed zero2 \
    --system 'prompt.txt' \
    --log_completions true \
    --dataloader_num_workers 52 \
    --dataset_num_proc 52 \
    >> train_logs.log 2>&1 &

```

![Image](https://github.com/user-attachments/assets/dae424d7-46bc-4024-837e-a489af0ca730)

![Image](https://github.com/user-attachments/assets/b9e4c193-c83e-4eb4-8d9b-6e075c9b5694)

**Comment by radna0** [comment]
Might @hjh0119 be able to help with this? Thanks in advance, the training adjusted for 14B works fine, on 32B it's keep being like this, I have tried changing configs multiple times, still no use

**Comment by hjh0119** [comment]
It looks like the model training has collapsed; perhaps adjusting the learning rate would help.

**Comment by radna0** [comment]
@hjh0119 I tried 1e-4, 1e-5 and even 1e-6, but there's no difference? This is happening on the distill 32B, the distill 14B works fine.

**Comment by radna0** [comment]
@hjh0119 1e-8 learning rate, and tried another run with repetition reward, but loss still goes to 0, grad_norm nan, gibberish outputs still.

## Repetition Penalty Function Problem
I have found out that the following function attributed to the model reward hacking, if completion is nothing or too short, the following function should return max_penalty

![Image](https://github.com/user-attachments/assets/169804af-6d1e-4531-bcc5-7cc56641dd90)
![Image](https://github.com/user-attachments/assets/9d7c46df-acf3-4103-9e09-1f3856dbdc62)

## Fix?  I'm still testing this
![Image](https://github.com/user-attachments/assets/7296b8f2-e10c-4923-81ec-0bb40bc21b85)

![Image](https://github.com/user-attachments/assets/28e2b278-033c-42ae-8df6-c9880be3c557)

=> The model seems to have learnt to do reward hacking and bypass learning at all, using correct repetition penalty fixed the following problem.

## Extras

@hjh0119 In the orginal paper there was a correctness discount factor, where can I modify the following params?

![Image](https://github.com/user-attachments/assets/48a37feb-83af-47b8-9598-efa04c281ba0)

**Comment by radna0** [comment]
Update: Model collapse still happened

**Comment by effortprogrammer** [comment]
Does anyone fix with this issue? I wonder if it came from deepspeed related issue

**Comment by hjh0119** [comment]
Feel free to reopen if you have any issues.
